### ISSUE TITLE
cannot override environment variable values in modules

### ISSUE THREAD

**Issue by rcernich** [cekit/cekit#283]
moduleA
```yaml
env:
- name: foo
  value: bar
```

moduleB
```yaml
env:
- name: foo
  value: baz
modules:
  install:
  - name: moduleA
```

image.yaml
```yaml
modules:
  install:
  - name: moduleB
```

Notice that image has foo=bar.

**Comment by dbecvarik** [comment]
IT seems something wrong is happening in merging. I'll probably take a deeper look at merging code as it was tweaked multiple times and its probably time to revisit it and see if it can be make less complex and less error prone.

**Comment by jmtd** [comment]
With the exact case above, using cekit built from develop as of around now, I get foo=baz in the generated Dockerfile. I.e. the parent module is not overridden by the child.

What behaviour do we want?

**Comment by dbecvarik** [comment]
parent values overrides child.. so if a module introduce any variable you can change it in image.yaml (image descriptor is parent of all modules)

**Comment by jmtd** [comment]
With 2.0.0rc4 I get foo=bar, with develop (`--version => 2.1.rc0`) I get foo=baz

**Comment by rcernich** [comment]
Here's a more specific example.  It's not a child module, but a subsequent module, so...

image.yaml
```yaml
modules:
  install:
  - name: moduleA
  - name: moduleB
```
Using the same module definitions listed above.

HTH

**Comment by rcernich** [comment]
[cekit-283.tar.gz](https://github.com/cekit/cekit/files/2274797/cekit-283.tar.gz)
Reproducer attached.  After building, the image should contain the following env variables:

- test1=from-moduleB
- test2=from-image

**Comment by jmtd** [comment]
>  Reproducer attached. After building, the image should contain the following env variables:

Should as in should in the working case?

My reading of that test case for the specific variable test1 is as follows

* not defined in image.yaml at all, fine
* image.yaml lists (in this order) modules C then B
* module C lists module A
* module A defines test2=from-moduleA, so we have our first value
* now we process module B: it first defines test2=from-moduleB, so that's the new value
* but now module B pulls in module A.

It does so *after* defining that variable itself, so I would actually expect test2 to be redefined again to from-moduleA. is that not anyone else's expectation?

**Comment by jmtd** [comment]
So current behaviour matches my expectation; @dbecvarik's work in #287 seems to match your expectation.

**Comment by dbecvarik** [comment]
@jmtd I'm trying to do something like reversed [Post-order Tree traversal](https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)) to serialize dependency graph this creates 'correct' order for merging/overriding descriptor. 
In such list of modules module 'A' values are overridden by module 'C' which is overridden by module 'e', etc...

**Comment by jmtd** [comment]
Ah I see so in Rob's example above, where module B defines test2 and pulls in module A, which defines test2; module B overrides module A. The order of env vars and modules in a particular YAML file is not relevant. Is that right? I forgot that yes, a module should override its children.

**Comment by dbecvarik** [comment]
@jmtd yes.. order in module is relevant only for `execute` block obviously :)

**Comment by rcernich** [comment]
@jmtd, I look at this pretty simply:

1. order the modules, in this case the order would be [ moduleA, moduleC, moduleB ]
2. apply them to the image in that order

The caveats to this are: 
1. module versions overridden in image.yaml, which would have the effect of simply changing the version applied, not necessarily the order in which they are applied
2. artifacts, which would override anything specified in any module
3. run, which would have to be merged as you went, since it should result in a single entry at the end of the generated file

**Comment by dbecvarik** [comment]
fixed by #287
